### PR TITLE
Enhance MDI support in LAMMPS in its MDI package

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

This PR does several things wrt support in LAMMPS for the MDI code coupling library.  All in package MDI.

LAMMPS can now be used in all of the modes MDI supports.  As a driver code, coupled to engines that are either stand-alone codes or plugin libraries.  As an engine code, either in stand-alone mode or as a plugin library.

a) Refactors the mdi/engine and fix mdi/engine commands, so that all the MDI logic is in the former.  Also changes
    the name of the command to mdi engine, so that other variant mdi commands can be added.
b) Adds a fix mdi/aimd command so that LAMMPS can act as a driver for an AIMD calculation.
c) Adds custom MDI engine commands appropriate for using LAMMPS as an MD engine
d) Adds a mdi plugin command so that LAMMPS as an MDI driver can use an engine either as a standalone code or a plugin library
e) Adds an examples/mdi dir with several use cases with many input scripts and a README to explain how to use LAMMPS with MDI in all the various modes

**Related Issue(s)**

N/A

**Author(s)**

Steve and Taylor Barnes (MolSSI at Virginia Tech)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Should just add functionality, but haven't yet tested for backward compatibility.

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by sjplimp** [comment]
@akohlmey two things we could use your help on - thanks!
(1) CMake support for building with new MDI package and its lib in lib/mdi
(2) lib/mdi/Install.py now downloads version 1.3.0 but I don't know how to do add the checksum

**Comment by akohlmey** [comment]
> @akohlmey two things we could use your help on - thanks! (1) CMake support for building with new MDI package and its lib in lib/mdi (2) lib/mdi/Install.py now downloads version 1.3.0 but I don't know how to do add the checksum

This is rather simple to do:
 - download the specific archive (or use the one that Install.sh downloads)
 - run: md5sum <name of tarball>
 - add/replace the version/md5sum in lib/mdi/Install.sh and cmake/Modules/Packages/MDI.cmake
 
please see commit a4018dbb4bb9f4968b5e8a6b6e882dc97f74790e with the correspondingly applied changes.

File: src/MDI/mdi_engine.cpp [hunk 1]
```diff
@@ -18,322 +18,1662 @@
 
 #include "mdi_engine.h"
 
+#include <limits>
```
Comment (akohlmey) [inline src/MDI/mdi_engine.cpp:21]: @sjplimp these changes undo updates to upstream that had aligned this code with the contributing guidelines about include file names and ordering.

File: src/MDI/mdi_engine.cpp [hunk 2]
```diff
@@ -18,322 +18,1662 @@
...
+  if (flag_create) create_system();
+  else {
+    if (flag_cell || flag_cell_displ) adjust_box();
+    if (flag_charges) adjust_charges();
+    if (flag_coords) adjust_coords();
+    if (flag_velocities) adjust_velocities();
+  }
 
-  // check that LAMMPS is setup as a compatible MDI engine
+<<<<<<< HEAD
```
Comment (akohlmey) [inline src/MDI/mdi_engine.cpp:587]: @sjplimp you have committed here an unresolved merge conflict. This causes the code to fail compiling.

**Comment by sjplimp** [comment]
@akohlmey I think this PR is now ready to merge.  One thing it may still lack is CMake support for building with the MDI library in lib/mdi.  There are a couple small code improvements I can make once Taylor implements a couple new features in MDI.  But the
current version of the PR works without them.  So I can make those changes later if you are ready to merge soon.

**Comment by akohlmey** [comment]
> @akohlmey I think this PR is now ready to merge. One thing it may still lack is CMake support for building with the MDI library in lib/mdi. There are a couple small code improvements I can make once Taylor implements a couple new features in MDI. But the current version of the PR works without them. So I can make those changes later if you are ready to merge soon.

This segfaults left and right when using the library interface.  nearly 90% of the unit tests are crashing because of that. https://ci.lammps.org/job/dev/job/pull_requests/job/ubuntu/job/unit_tests/job/cmake_mpi_openmp_bigbig_shared/4419/#showFailuresLink

Also, please see: https://ci.lammps.org/job/dev/job/pull_requests/job/checkstyle/4946/
This is not a "hard" error but should be addressed regardless (cd src; make fix-whitespace; git add ... ; git commit ; git push)

**Review by akohlmey — APPROVED** [review]

**Pull request merged**